dev_docs/tutorials/versioning_interfaces.mdx

```diff
@@ -0,0 +1,442 @@
+```
+
+How we use these types depends on whether the code is aware of versions or not. Generally, code that is aware of multiple versions will lead to greater complexity. We might have a logger function that expects the latest dashboards:
+
+```ts
+// This type is being pulled from "latest", so whenever a new latest is linked,
+// it will also point to the latest WeatherDashboard.
+import type { WeatherDashboard } from '../common';
+
+function logIt(dashboard: WeatherDashboard) {
+  logger.debug(dashboard);
+}
+```
+
+Or we might have a route handler that expects a specific version our domain objects:
+
+```ts
+// Now we bring it all together by using our versioned types in our handler from before:
+import type { weatherDashboardV1 } from '../common';
+
+const handler = (ctx, req, res) => {
+  const body: weatherDashboardV1.GetWeatherDashboardHTTPResponse = { ... };
+  return res.ok({ body });
+}
+```
+
+<DocCallOut title="Minimize the impact of refactors">
+  By referencing types linked from `latest.ts` we will not have to refactor all our app code to point to the latest version of `WeatherDashboard`. This happens by simply re-exporting the newest types from `latest.ts`. Otherwise, code that needs to know about a specific version or versions will have access to it by using the appropriate namespace.
+</DocCallOut>
+
+### Introducing a new field to `DataSource`
+
+Now that we have prepared our interfaces to be versioned, let's introduce the `name` field to `DataSource`.
+
+Create a new `v2.ts` file under the `data_source` directory:
+
+```
+common
+  data_source
+    v1.ts
+    v2.ts
+```
+
+The `v2.ts` file will contain:
+
+```ts
+interface DataSource {
+  /** Some URL that provides weather data in an expected format for scraping */
+  url: string;
+  /** A user-friendly name */
+  name: string;
+  /** A longer description of this data source */
+  description?: string;
+  /** The data view data is stored in */
+  dataViewId: string;
+  /** Username to use with source URL */
+  username?: string;
+  /** Password to use with source URL */
+  password?: string;
+}
+```
+
+<DocCallOut title="What is a breaking change?">
+  If we had only added the optional `description` field to `DataSource` this would be a backwards-compatible change that does not require a new version. We could have remained on `v1` of the `DataSource` interface.
+</DocCallOut>
+
+We also need to update the `WeatherDashboard` interface to use our new `DataSource` interface:
+
+```
+common
+  weather_dashboard
+    v1.ts
+    v2.ts
+```
+
+```ts
+// common/weather_dashboard/v2.ts
+
+// This is largely a copy of v1.ts, but using a new WeatherDashboard type in all
+// the relevant places
+
+/** Fictional type, only used for this example as an external type */
+import type { PortableDashboard } from '@kbn/dashboard/common';
+import type { DataSource } from '../data_source/v2';
+
+// Optionally, re-export the entire set of types.
+export * from './v1';
```

Comment:
Would the re-export be overriden by any interface below it with the same name?

Reply:
Yes, exactly that